src/lib/Lower.hs

```diff
@@ -0,0 +1,130 @@
+-- Copyright 2022 Google LLC
+--
+-- Use of this source code is governed by a BSD-style
+-- license that can be found in the LICENSE file or at
+-- https://developers.google.com/open-source/licenses/bsd
+
+module Lower (lowerFullySequential) where
+
+import Data.Map.Strict qualified as M
+import Control.Monad
+import Unsafe.Coerce
+
+import Types.Core
+import Types.Primitives
+
+import Name
+import MTL1
+import Core
+import Builder
+import QueryType
+import GenericTraversal
+
+lowerFullySequential :: EnvReader m => Block n -> m n (Block n)
+lowerFullySequential b = liftM fst $ liftGenericTraverserM LFS $ traverseGenericE b
+
+data LFS (n::S) = LFS
+type LowerM = GenericTraverserM LFS
+instance SinkableE LFS where
+  sinkingProofE _ LFS = LFS
+instance HoistableState LFS where
+  hoistState _ _ LFS = LFS
+  {-# INLINE hoistState #-}
+
+instance GenericTraverser LFS where
+  traverseExpr expr = case expr of
+    Hof (For dir ixTy (Lam body)) -> traverseFor Nothing dir ixTy body
```

Comment:
Can there be any `For`s whose body is not inlined as a `Lam`?  Do we want to skip those, or assert they are impossible?

Reply:
All `For`s always embed a `Lam`. It's the only way to elaborate them from surface syntax, and we never outline functions, so it always holds.